util/logtools/encounter_tools.ts

```diff
@@ -0,0 +1,313 @@
+  endLine?: string;
+  startTime?: Date;
+  endTime?: Date;
+};
+
+type FightEncInfo = {
+  name?: string;
+  zoneId?: number;
+  startLine?: string;
+  endLine?: string;
+  startTime?: Date;
+  endTime?: Date;
+  endType?: string;
+  sealName?: string;
+};
+
+export class EncounterFinder {
+  currentZone: ZoneEncInfo;
+  currentFight: FightEncInfo;
+  currentSeal: string | null;
+  zoneInfo: {
+    readonly exVersion: number;
+    readonly contentType?: number;
+    readonly name: LocaleText;
+    readonly offsetX: number;
+    readonly offsetY: number;
+    readonly sizeFactor: number;
+    readonly weatherRate: number;
+  } | null;
+
+  haveWon: boolean;
+  haveSeenSeals: boolean;
+
+  regex: {
+    changeZone: CactbotBaseRegExp<'ChangeZone'>;
+    cactbotWipe: CactbotBaseRegExp<'GameLog'>;
+    win: CactbotBaseRegExp<'ActorControl'>;
+    wipe: CactbotBaseRegExp<'ActorControl'>;
+    commence: CactbotBaseRegExp<'ActorControl'>;
+    playerAttackingMob: CactbotBaseRegExp<'Ability'>;
+    mobAttackingPlayer: CactbotBaseRegExp<'Ability'>;
+  };
+
+  sealRegexes: Array<CactbotBaseRegExp<'GameLog'>>;
+  unsealRegexes: Array<CactbotBaseRegExp<'GameLog'>>;
+
+  initializeZone(): void {
+    this.currentZone = {};
+  }
+  initializeFight(): void {
+    this.currentFight = {};
+  }
+  constructor() {
+    this.currentZone = {};
+    this.currentFight = {};
+    this.currentSeal = null;
+    // May be null for some zones.
+    this.zoneInfo = null;
+
+    this.haveWon = false;
+    this.haveSeenSeals = false;
+
+    this.regex = {
+      changeZone: NetRegexes.changeZone({ id: '*' }),
+      cactbotWipe: NetRegexes.echo({ line: 'cactbot wipe.*?' }),
+      win: NetRegexes.network6d({ command: '40000003' }),
+      wipe: NetRegexes.network6d({ command: '40000010' }),
+      commence: NetRegexes.network6d({ command: '4000000[16]' }),
+      playerAttackingMob: NetRegexes.ability({ sourceId: '1.{7}', targetId: '4.{7}' }),
+      mobAttackingPlayer: NetRegexes.ability({ sourceId: '4.{7}', targetId: '1.{7}' }),
+    };
+
+    // TODO: consider also doing this with lang? But lang only deals with Regexes???
+    // Sorry, this is quite a bit of a hack.
+    this.sealRegexes = [];
+    this.unsealRegexes = [];
+
+    const sealReplace = commonReplacement.replaceSync[syncKeys.seal];
+    if (!sealReplace)
+      throw new Error('missing seal regex');
+    for (const regexString of Object.values(sealReplace)) {
+      const line = `${regexString?.replace('$1', '(?<seal>.*?)')}.*?`;
+      this.sealRegexes.push(NetRegexes.message({ line: line }));
+    }
+
+    const unsealReplace = commonReplacement.replaceSync[syncKeys.unseal];
+    if (!unsealReplace)
+      throw new Error('missing unseal regex');
+    for (const lang of Object.values(unsealReplace)) {
+      const line = `*.?${lang}.*?`;
+      this.unsealRegexes.push(NetRegexes.message({ line: line }));
+    }
+  }
+
+  skipZone(): boolean {
+    if (!this?.currentZone?.zoneId || this.zoneInfo === null)
```

Comment:
No need for `this?` here and elsewhere.

```suggestion
    if (!this.currentZone?.zoneId || this.zoneInfo === null)
```